src/linked_list.js

```diff
@@ -0,0 +1,136 @@
+'use strict'
+
+function Node( data ) {
+  this._data = data
+  this._next = null
+}
+
+export default class LinkedList {
+  constructor() {
+    this._tail = null
+    this._head = null
+    this._size = 0
+  }
+
+  size() {
+    return this._size
+  }
+
+  isEmpty() {
+    return this._size == 0
+  }
+
+  clear() {
+    this._tail = null
+    this._head = null
+    this._size = 0
+  }
+
+  getHeadNode() {
+    if ( this._head ) { return this._head }
+    else { return null }
+  }
+
+  getTailNode() {
+    if ( this._tail ) { return this._tail }
+    else { return null }
+  }
+
+  insert( data ) {
+    const newNode = new Node( data )
+    if ( this._size === 0 ) {
+      this._tail = newNode
+    } else {
+      newNode._next = this._head
+    }
+    this._head = newNode
+    ++this._size
+  }
+
+  insertFirst( data ) {
+    const newNode = new Node( data )
+    if ( this._size === 0 ) {
+      this._head = newNode
+    } else {
+      const oldTail = this._tail
+      oldTail._next = newNode
+    }
+    this._tail = newNode
+    ++this._size
+  }
+
+  remove() {
+    if ( this.isEmpty() ) { return null }
+    const oldHead = this._head
+    --this._size
+    let newHead = oldHead.next
+    this._head = newHead
+    return oldHead._data
+  }
+
+  removeFirst() {
+    if ( this.isEmpty() ) { return null }
+    const oldTail = this._tail
+    --this._size
+    let newFront = this._head
+    for ( const i = 1; i < this._size; ++i) {
+      newFront = newFront._next
+    }
+    newFront._next = null
+    this._tail = newFront
+    return oldTail._data
+  }
+
+  find( query ) {
+    if ( this.isEmpty() ) { return -1 }
+    let iteratorNode = this.getHeadNode()
+    while ( iteratorNode != null ) {
+      if ( iteratorNode._data == query ) {
+        return iteratorNode }
+      else { iteratorNode = iteratorNode._next }
+    }
+    if ( iteratorNode === null ) { return -1 }
+  }
+
+  contains( query ) {
+    if ( this.isEmpty() ) { return false }
+    let iteratorNode = this.getHeadNode()
+    while ( iteratorNode != null ) {
+      if ( iteratorNode._data == query ) {
+        return true }
+      else { iteratorNode = iteratorNode._next }
+    }
+    return false
+  }
+
+  insertBefore( query, newData ) {
```

Comment:
I'm trying to convince myself that this implementation works, but I'm not seeing it.  Let's say we have a linked list consisting of the following structure:
```
1 ---> 2 ---> 3 ---> null
```
I then place a call `insertBefore( 3, 'A' )`.  Tracing your implementation, `targetNode` would be a reference to the `Node` containing 3 after the line `const targetNode = this.find( query )`:
```
1 ---> 2 ---> 3 ---> null
              t
```
`targetNode` (`t` in the picture above), is not -1, so `newNode` is instantiated:
```
newNode ---> 'A' ---> null

1 ---> 2 ---> 3 ---> null
              t
```
Now, we execute the line `newNode._next = targetNode._next`, changing the picture to (note that the picture hasn't changed - `t.next` is `null`, and so we're just reassigning `null` to `newNode.next`:
```
newNode ---> 'A' ---> null

1 ---> 2 ---> 3 ---> null
              t
```
Now we reset `tail` to point at `newNode`, then set `targetNode`'s `next`, making the list look like this:
```
                      n & tail
1 ---> 2 ---> 3 ---> 'A' ---> null
              t
```
All this to say that I think this is `insertAfter` rather than `insertBefore`?